// @framer/kocijan.linear-progress-bar/code/linearProgressBar.tsx
import * as React from "react"
import { PropertyControls, ControlType } from "framer"
import styled from "styled-components"

// Create a Wrapper component that'll render a <section> tag with some styles

const Wrapper = styled.div``

const linearProgressBar = styled.div``

// Define type of property
interface Props {
    progressBarbackgroundColor: string
    progressBarTickness: number
    progressBar: number
    fontColor: string
    fontSize: number
    hideText: "visible" | "hidden"
    backgroundColor: string
    fontWeight: number
    borderRadius: number
}

export class linearProggBar extends React.Component<Props> {
    // Set default properties
    static defaultProps = {
        progressBarbackgroundColor: "#0099CC",
        progressBarTickness: "20",
        progressBar: "20",
        fontColor: "#ffffff",
        hideText: "visible",
        fontSize: "12",
        backgroundColor: "#F9F9F9",
        fontWeight: "600",
        borderRadius: "4",
    }

    // Items shown in property panel
    static propertyControls: PropertyControls = {
        progressBarbackgroundColor: {
            type: ControlType.Color,
            title: "Progress Color",
            defaultValue: "green",
        },
        
        backgroundColor: {
            type: ControlType.Color,
            title: "Bckg Color",
            defaultValue: "lightGray",
        },
        borderRadius: {
            type: ControlType.Number,
            title: "Radius",
        },

        progressBar: {
            type: ControlType.Number,
            title: "Set %",
            min: 0,
            max: 100,
            defaultValue: 48,
        },
        progressBarTickness: {
            type: ControlType.Number,
            title: "Thickness",
            defaultValue: 20,
        },

// set the text as property in bar        
        hideText: {
            type: ControlType.SegmentedEnum,
            options: ["visible", "hidden"],
            optionTitles: ["Visible", "Hidden"],
            title: "Text %",
            defaultValue: "hidden",
        },

        fontColor: {
            type: ControlType.Color,
            title: "Color",
            defaultValue: "white",
            hidden: (props) => props.hideText !== "visible",
        },

        fontSize: {
            type: ControlType.Number,
            title: "Size",
            defaultValue: 11,
            hidden: (props) => props.hideText !== "visible",
        },
        fontWeight: {
            type: ControlType.Number,
            title: "Weight",
            min: 100,
            max: 1000,
            step: 100,
            defaultValue: 500,
            hidden: (props) => props.hideText !== "visible",
        },
    }

    render() {
        return (
            <Wrapper
                style={{
                    ...wrapStyle,
                    ...{ height: this.props.progressBarTickness },
                    ...{ backgroundColor: this.props.backgroundColor },
                    ...{ borderRadius: `${this.props.borderRadius}px` },
                }}
            >
                <div
                    style={{
                        ...progressBarStyle,
                        ...{ width: `${this.props.progressBar}%` },
                        ...{ height: this.props.progressBarTickness },
                        ...{
                            backgroundColor: this.props
                                .progressBarbackgroundColor,
                        },
                    }}
                >
                    <span
                        style={{
                            ...fontStyle,
                            ...{ color: this.props.fontColor },
                            ...{ fontSize: this.props.fontSize },
                            ...{ visibility: this.props.hideText },
                            ...{ fontWeight: this.props.fontWeight },
                        }}
                    >
                        {this.props.progressBar}%
                    </span>
                </div>
            </Wrapper>
        )
    }
}

const wrapStyle: React.CSSProperties = {
    display: "flex",
    overflow: "hidden",
    transition: "1.4s cubic-bezier(0.2, 0.8, 0.2, 1)",
}

const fontStyle: React.CSSProperties = {
    fontWeight: 600,
}

const progressBarStyle: React.CSSProperties = {
    display: "flex",
    flexDirection: "row",
    overflow: "hidden",
    justifyContent: "center",
    alignItems: "center",
    transition: "1.4s cubic-bezier(0.2, 0.8, 0.2, 1)",
}
