// @framer/talkaboutdesign.design-hunt/code/Design_Hunt.tsx
import * as React from "react"
import { PropertyControls, ControlType, Animatable } from "framer"

// Define type of property
interface Props {
    fill: string
    url: string
    width: number
    height: number
    radius: number
}

export class Design_Hunt extends React.Component<Props> {
    static propertyControls: PropertyControls<Props> = {
        fill: {
            type: ControlType.Enum,
            title: "Background",
            options: ["cover", "contain", "fill"],
        },
        radius: { type: ControlType.Number, min: 0, max: 200, title: "Radius" },
    }

    static defaultProps = {
        url: "https://designhunt.us/images/background_graphic.png",
        radius: 0,
        fill: "cover",
    }

    state = { url: "https://designhunt.us/images/background_graphic.png" }

    componentDidMount() {
        this.setImage()
    }

    async setImage() {
        const randomImageUrl = "https://designhunt.us/api/feed/randomImage"
        const url = `${randomImageUrl}`
        const result = await fetch(url)
        const data = await result.json()
        this.setState({ url: data[0] })
    }

    renderFill() {
        return (
            <div
                style={{
                    width: this.props.width,
                    height: this.props.height,
                    backgroundImage: `url(${this.state.url})`,
                    backgroundSize: "100% 100%",
                    backgroundPosition: "center top",
                    backgroundRepeat: "no-repeat",
                    display: "flex",
                    justifyContent: "center",
                    alignItems: "center",
                    overflow: "hidden",
                    position: "relative",
                    borderRadius: `${this.props.radius / 2}%`,
                }}
            />
        )
    }

    renderRegular() {
        return (
            <div
                style={{
                    width: this.props.width,
                    height: this.props.height,
                    backgroundImage: `url(${this.state.url})`,
                    backgroundSize: this.props.fill,
                    backgroundPosition: "center center",
                    backgroundRepeat: "no-repeat",
                    display: "flex",
                    justifyContent: "center",
                    alignItems: "center",
                    overflow: "hidden",
                    position: "relative",
                    borderRadius: `${this.props.radius / 2}%`,
                }}
            />
        )
    }

    render() {
        if (this.props.fill === "fill") {
            return this.renderFill()
        } else {
            return this.renderRegular()
        }
    }
}
