// @framer/ruinunes.news-feed/code/News_stack.tsx
import * as React from "react"
import { addPropertyControls, ControlType, Stack } from "framer"
import { Article } from "./canvas"

// Open Preview (CMD + P)
// API Reference: https://www.framer.com/api

let data = []

export function News_stack(props) {
    const url =
        "https://newsapi.org/v2/everything?q=" +
        props.keyword +
        "&domains=" +
        props.sources +
        "&apiKey=" +
        props.apiKey

    fetch(url).then(response =>
        response.json().then(response => {
            data = response.articles
            console.log(data)
        })
    )

    return (
        <Stack
            gap={props.gap}
            distribution="start"
            alignment="start"
            direction="vertical"
            width="100%"
            height="100%"
        >
            {data ? (
                data
                    .slice(0, props.articles)
                    .map(data => (
                        <Article
                            width="100%"
                            articleSource={data.source.name}
                            articleTitle={data.title}
                            articleImage={data.urlToImage}
                        />
                    ))
            ) : (
                <Article
                    width="100%"
                    articleSource="News API"
                    articleTitle="Get your news api key here: https://newsapi.org"
                    articleImage="https://images.unsplash.com/photo-1541560052-5e137f229371?ixlib=rb-1.2.1ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1234&q=80"
                />
            )}
        </Stack>
    )
}

News_stack.defaultProps = {
    api: "Get API key here: https://newsapi.org",
    apiKey: "",
    keyword: "framer",
    sources: "framer.com",
    gap: 8,
    articles: 4,
}

addPropertyControls(News_stack, {
    api: { type: ControlType.String, title: "API" },
    apiKey: { type: ControlType.String, title: "Key" },
    keyword: { type: ControlType.String, title: "Keywork" },
    sources: { type: ControlType.String, title: "Sources" },
    articles: {
        type: ControlType.Number,
        title: "# Articles",
        min: 1,
        max: 20,
    },
    gap: { type: ControlType.Number, title: "Gap", max: 100 },
})
